lib/spack/spack/spec.py

```diff
@@ -766,13 +767,133 @@ def __str__(self):
             for key in sorted_keys) + cond_symbol
 
 
-class DependencyMap(lang.HashableMap):
-    """Each spec has a DependencyMap containing specs for its dependencies.
-       The DependencyMap is keyed by name. """
+def _sort_by_dep_types(dspec):
+    # Use negation since False < True for sorting
+    return tuple(t not in dspec.deptypes for t in ("link", "run", "build", "test"))
+
+
+#: Enum for edge directions
+EdgeDirection = lang.enum(parent=0, child=1)
+
+
+@lang.lazy_lexicographic_ordering
+class _EdgeMap(Mapping):
+    """Represent a collection of edges (DependencySpec objects) in the DAG.
+
+    Objects of this class are used in Specs to track edges that are
+    outgoing towards direct dependencies, or edges that are incoming
+    from direct dependents.
+
+    Edges are stored in a dictionary and keyed by package name.
+    """
+    def __init__(self, store_by=EdgeDirection.child):
+        # Sanitize input arguments
+        msg = 'unexpected value for "store_by" argument'
+        assert store_by in (EdgeDirection.child, EdgeDirection.parent), msg
+
+        #: This dictionary maps a package name to a list of edges
+        #: i.e. to a list of DependencySpec objects
+        self.edges = {}
+        self.store_by_child = (store_by == EdgeDirection.child)
+
+    def __getitem__(self, key):
+        return self.edges[key]
+
+    def __iter__(self):
+        return iter(self.edges)
+
+    def __len__(self):
+        return len(self.edges)
+
+    def add(self, edge):
+        """Adds a new edge to this object.
+
+        Args:
+            edge (DependencySpec): edge to be added
+        """
+        key = edge.spec.name if self.store_by_child else edge.parent.name
+
+        current_list = self.edges.setdefault(key, [])
+
+        # The list of edges is empty, just append the edge
+        if not current_list:
+            current_list.append(edge)
+            return
+
+        # If there's something in the list, check if we need to update an
+        # already existing entry.
+        for dep in current_list:
+            if edge.spec == dep.spec and edge.parent == dep.parent:
```

Comment:
Jut as a preliminary look into this -- I think the solution here is to *not* do this duplicate checking (at least not with a full spec compare) on insert.

The database never adds duplicates, and the CLI spec parser doesn't allow it.  When I do a `buildcache list`, it looks like the vast majority of adds are not duplicates, but there are a few (29 out of tens of thousands) that are.  In all cases they aren't adding new deptypes or anything -- they're adding the same deptypes that are already there.  I'm going to look at usages of `add_dependency_edge` and `_add_dependency` to see if we can eliminate this checking.

Reply:
If we just store a separate edge for each dep type between the two specs, will that hurt us anywhere else? We'll have to change how we check the edge list for dependencies, but that should be doable.

And then we can just check name and type here, and if the name/type combination already exists, we can raise.

Reply:
That doesn’t work for dependents. You can have the same name/type multiple times.

Reply:
There's also the fact that here we are really talking about edge attributes. Right now we only have dependency types but in the future there will be more (virtuals provided, use variants etc.) - so one edge per attribute value I think won't work.

Reply:
Leaving a note on the idea to use `id` instead of `__eq__` for this check. When we compare abstract specs using the check above with `__eq__` we have:
```python
Spack version 0.17.1
Python 3.8.10, Linux x86_64
>>> import spack.spec
>>> s = spack.spec.Spec('root')
>>> d1 = spack.spec.Spec('child@1.0')
>>> d2 = spack.spec.Spec('child@1.0')
>>> s.add_dependency_edge(d1, deptype=('link',))
>>> s.add_dependency_edge(d2, deptype=('build',))
>>> str(s)
'root ^child@1.0'
```
i.e. the two calls unify d1 and d2 into a single `(build,link)` edge. This seems sensible to me. Using `id` will surely be faster, but opens for situations like the following on abstract specs:
```python
>>> import spack.spec
>>> s = spack.spec.Spec('root')
>>> d1 = spack.spec.Spec('child@1.0')
>>> d2 = spack.spec.Spec('child@1.0')
>>> s.add_dependency_edge(d1, deptype=('link',))
>>> s.add_dependency_edge(d2, deptype=('link',))
>>> str(s)
'root ^child@1.0 ^child@1.0'
```
where adding even the same dependency type result in a duplicate edge because `d1` and `d2` are not the same object.

Reply:
Considering that `id(s1) == id(s2)` implies both:
- `s1 == s2`
- `s1.*_hash() == s2.*_hash()`, if the specs are concrete

is it an idea we can consider to keep them both and compare by `id` first (fast method) and by hash/equality later, i.e. if `id` are not the same?

Reply:
It seems the idea above is already implemented, so taking note of that here:

https://github.com/spack/spack/blob/6934703ffbae9b475b66c6321a1fb06c500105fa/lib/spack/llnl/util/lang.py#L436-L439

After a few DMs exchange with @tgamblin the idea is:
1. Remove the double comparison from line 826
2. Have `Spec.add_dependency` (see https://github.com/spack/spack/pull/28673#discussion_r820165922) take care or vetting input specs if needed

Reply:
See df30e6c 085dcf8 and 43bd292